src/part_file.cpp

```diff
@@ -210,24 +213,24 @@ namespace libtorrent
 			return -1;
 		}
 
-		int slot = i->second;
+		int const slot = i->second;
 
 		open_file(file::read_write | file::attribute_hidden, ec);
 		if (ec) return -1;
 
 		l.unlock();
 
-		boost::int64_t slot_offset = boost::int64_t(m_header_size) + boost::int64_t(slot) * m_piece_size;
+		boost::int64_t const slot_offset = calc_offset(slot, m_header_size, m_piece_size);
 		return m_file.readv(slot_offset + offset, bufs, num_bufs, ec);
 	}
 
 	void part_file::open_file(int mode, error_code& ec)
 	{
-		if (m_file.is_open()
-			&& ((m_file.open_mode() & file::rw_mask) == mode
-				|| mode == file::read_only)) return;
+		if (m_file.is_open() && mode == file::read_only
+			|| (m_file.open_mode() & file::rw_mask) == (mode & file::rw_mask))
+			return;
```

Comment:
Here, wrong `mode` argument comparison caused file to be re-opened most of time.